File: app/store/event_store.py
from __future__ import annotations

import json
from datetime import datetime

from psycopg.rows import dict_row

from app.models import NormalizedEvent, ScoredEvent
from app.rules.weights import ALL_SECTORS
from app.store.db import get_db


def reset_scored_data() -> None:
    conn = get_db()
    cur = conn.cursor()
    cur.execute("TRUNCATE scored_events")
    cur.execute("TRUNCATE normalized_events")
    conn.commit()
    conn.close()


def save_normalized(event: NormalizedEvent) -> None:
    conn = get_db()
    cur = conn.cursor()
    cur.execute(
        """
        INSERT INTO normalized_events
        (raw_event_id, event_type, policy_domain, risk_signal, rate_signal, geo_signal, sector_impacts, sentiment, rationale, channels, confidence, regime, baseline)
        VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
        ON CONFLICT (raw_event_id) DO UPDATE SET
            event_type = EXCLUDED.event_type,
            policy_domain = EXCLUDED.policy_domain,
            risk_signal = EXCLUDED.risk_signal,
            rate_signal = EXCLUDED.rate_signal,
            geo_signal = EXCLUDED.geo_signal,
            sector_impacts = EXCLUDED.sector_impacts,
            sentiment = EXCLUDED.sentiment,
            rationale = EXCLUDED.rationale,
            channels = EXCLUDED.channels,
            confidence = EXCLUDED.confidence,
            regime = EXCLUDED.regime,
            baseline = EXCLUDED.baseline
        """,
        (
            event.raw_event_id,
            event.event_type,
            event.policy_domain,
            event.risk_signal,
            event.rate_signal,
            event.geo_signal,
            json.dumps(event.sector_impacts, ensure_ascii=True),
            event.sentiment,
            event.rationale,
            json.dumps(event.channels, ensure_ascii=True),
            event.confidence,
            json.dumps(event.regime, ensure_ascii=True),
            json.dumps(event.baseline, ensure_ascii=True),
        ),
    )
    conn.commit()
    conn.close()


def fetch_unscored_events(limit: int = 200) -> list[NormalizedEvent]:
    conn = get_db()
    cur = conn.cursor(row_factory=dict_row)
    cur.execute(
        """
        SELECT n.* FROM normalized_events n
        LEFT JOIN scored_events s ON s.raw_event_id::text = n.raw_event_id::text
        WHERE s.raw_event_id IS NULL
        ORDER BY n.raw_event_id DESC
        LIMIT %s
        """,
        (limit,),
    )
    rows = cur.fetchall()
    conn.close()

    events: list[NormalizedEvent] = []
    for row in rows:
        events.append(
            NormalizedEvent(
                raw_event_id=row["raw_event_id"],
                event_type=row["event_type"],
                policy_domain=row["policy_domain"],
                risk_signal=row["risk_signal"],
                rate_signal=row["rate_signal"],
                geo_signal=row["geo_signal"],
                sector_impacts=row["sector_impacts"],
                sentiment=row["sentiment"],
                rationale=row["rationale"],
                channels=row.get("channels") or [],
                confidence=row.get("confidence", 0.6),
                regime=row.get("regime") or {},
                baseline=row.get("baseline") or {},
            )
        )
    return events


def fetch_normalized_event(raw_event_id: str) -> NormalizedEvent | None:
    conn = get_db()
    cur = conn.cursor(row_factory=dict_row)
    cur.execute("SELECT * FROM normalized_events WHERE raw_event_id = %s", (raw_event_id,))
    row = cur.fetchone()
    conn.close()
    if not row:
        return None
    return NormalizedEvent(
        raw_event_id=row["raw_event_id"],
        event_type=row["event_type"],
        policy_domain=row["policy_domain"],
        risk_signal=row["risk_signal"],
        rate_signal=row["rate_signal"],
        geo_signal=row["geo_signal"],
        sector_impacts=row["sector_impacts"],
        sentiment=row["sentiment"],
        rationale=row["rationale"],
        channels=row.get("channels") or [],
        confidence=row.get("confidence", 0.6),
        regime=row.get("regime") or {},
        baseline=row.get("baseline") or {},
    )


def save_scored(event: ScoredEvent) -> None:
    conn = get_db()
    cur = conn.cursor()
    cur.execute(
        """
        INSERT INTO scored_events
        (raw_event_id, event_type, policy_domain, risk_signal, rate_signal, geo_signal, sector_impacts, sentiment, rationale,
         fx_state, sector_scores, total_score, created_at, channels, confidence, regime, baseline)
        VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
        ON CONFLICT (raw_event_id) DO UPDATE SET
            event_type = EXCLUDED.event_type,
            policy_domain = EXCLUDED.policy_domain,
            risk_signal = EXCLUDED.risk_signal,
            rate_signal = EXCLUDED.rate_signal,
            geo_signal = EXCLUDED.geo_signal,
            sector_impacts = EXCLUDED.sector_impacts,
            sentiment = EXCLUDED.sentiment,
            rationale = EXCLUDED.rationale,
            fx_state = EXCLUDED.fx_state,
            sector_scores = EXCLUDED.sector_scores,
            total_score = EXCLUDED.total_score,
            created_at = EXCLUDED.created_at,
            channels = EXCLUDED.channels,
            confidence = EXCLUDED.confidence,
            regime = EXCLUDED.regime,
            baseline = EXCLUDED.baseline
        """,
        (
            event.raw_event_id,
            event.event_type,
            event.policy_domain,
            event.risk_signal,
            event.rate_signal,
            event.geo_signal,
            json.dumps(event.sector_impacts, ensure_ascii=True),
            event.sentiment,
            event.rationale,
            event.fx_state,
            json.dumps(event.sector_scores, ensure_ascii=True),
            event.total_score,
            event.created_at,
            json.dumps(event.channels, ensure_ascii=True),
            event.confidence,
            json.dumps(event.regime, ensure_ascii=True),
            json.dumps(event.baseline, ensure_ascii=True),
        ),
    )
    conn.commit()
    conn.close()


def fetch_scored_event(raw_event_id: str) -> ScoredEvent | None:
    conn = get_db()
    cur = conn.cursor(row_factory=dict_row)
    cur.execute("SELECT * FROM scored_events WHERE raw_event_id = %s", (raw_event_id,))
    row = cur.fetchone()
    conn.close()
    if not row:
        return None
    return ScoredEvent(
        raw_event_id=row["raw_event_id"],
        event_type=row["event_type"],
        policy_domain=row["policy_domain"],
        risk_signal=row["risk_signal"],
        rate_signal=row["rate_signal"],
        geo_signal=row["geo_signal"],
        sector_impacts=row["sector_impacts"],
        sentiment=row["sentiment"],
        rationale=row["rationale"],
        fx_state=row["fx_state"],
        sector_scores=row["sector_scores"],
        total_score=row["total_score"],
        created_at=row["created_at"],
        channels=row.get("channels") or [],
        confidence=row.get("confidence", 0.6),
        regime=row.get("regime") or {},
        baseline=row.get("baseline") or {},
    )


def list_timeline(limit: int = 50) -> list[dict[str, object]]:
    conn = get_db()
    cur = conn.cursor(row_factory=dict_row)
    cur.execute(
        """
        SELECT r.title, r.url, r.published_at, r.sector, s.risk_signal, s.rate_signal, s.geo_signal, s.fx_state, s.sentiment, s.total_score
        FROM raw_events r
        JOIN scored_events s ON s.raw_event_id::text = r.id
        ORDER BY r.published_at DESC
        LIMIT %s
        """,
        (limit,),
    )
    rows = cur.fetchall()
    conn.close()

    return [
        {
            "title": row["title"],
            "url": row["url"],
            "published_at": row["published_at"].isoformat()
            if hasattr(row["published_at"], "isoformat")
            else row["published_at"],
            "sector": row["sector"],
            "risk_signal": row["risk_signal"],
            "rate_signal": row["rate_signal"],
            "geo_signal": row["geo_signal"],
            "fx_state": row["fx_state"],
            "sentiment": row["sentiment"],
            "total_score": row["total_score"],
        }
        for row in rows
    ]


def sector_heatmap() -> dict[str, float]:
    conn = get_db()
    cur = conn.cursor(row_factory=dict_row)
    cur.execute("SELECT sector_scores FROM scored_events")
    rows = cur.fetchall()
    conn.close()

    totals: dict[str, float] = {}
    for row in rows:
        scores = row["sector_scores"]
        for sector, score in scores.items():
            totals[sector] = totals.get(sector, 0.0) + float(score)
    for sector in ALL_SECTORS:
        totals.setdefault(sector, 0.0)
    return {sector: round(value, 3) for sector, value in totals.items()}


def graph_edges(limit: int = 100) -> list[dict[str, object]]:
    conn = get_db()
    cur = conn.cursor(row_factory=dict_row)
    cur.execute(
        """
        SELECT r.title, s.fx_state, s.risk_signal, s.rate_signal, s.geo_signal, s.sector_scores
        FROM raw_events r
        JOIN scored_events s ON s.raw_event_id::text = r.id
        ORDER BY r.published_at DESC
        LIMIT %s
        """,
        (limit,),
    )
    rows = cur.fetchall()
    conn.close()

    edges = []
    for row in rows:
        scores = row["sector_scores"]
        for sector, score in scores.items():
            edges.append(
                {
                    "event": row["title"],
                    "fx": row["fx_state"],
                    "risk_signal": row["risk_signal"],
                    "rate_signal": row["rate_signal"],
                    "geo_signal": row["geo_signal"],
                    "sector": sector,
                    "weight": float(score),
                    "fx_theme": row["fx_state"],
                }
            )
    if not edges:
        return [
            {
                "event": "Sample event (default)",
                "fx": "USD:+0 JPY:+0 EUR:+0 EM:+0",
                "risk_signal": "neutral",
                "rate_signal": "none",
                "geo_signal": "none",
                "sector": "Energy",
                "weight": 1,
                "fx_theme": "neutral",
            }
        ]
    return edges


def latest_created_at() -> datetime | None:
    conn = get_db()
    cur = conn.cursor(row_factory=dict_row)
    cur.execute("SELECT created_at FROM scored_events ORDER BY created_at DESC LIMIT 1")
    row = cur.fetchone()
    conn.close()
    if not row:
        return None
    created_at = row["created_at"]
    return created_at if isinstance(created_at, datetime) else datetime.fromisoformat(created_at)

File: server/src/main/java/com/finintel/app/store/EventStore.java
package com.finintel.app.store;

import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.stereotype.Repository;

import com.finintel.app.model.NormalizedEvent;
import com.finintel.app.model.RawEvent;
import com.finintel.app.model.ScoredEvent;
import com.finintel.app.rules.Weights;
import com.finintel.app.util.JsonCodec;

@Repository
public class EventStore {
    private final JdbcTemplate jdbcTemplate;
    private final JsonCodec json;

    public EventStore(JdbcTemplate jdbcTemplate, JsonCodec json) {
        this.jdbcTemplate = jdbcTemplate;
        this.json = json;
    }

    public void resetScoredData() {
        jdbcTemplate.execute("TRUNCATE scored_events");
        jdbcTemplate.execute("TRUNCATE normalized_events");
    }

    public int saveRawEvents(Iterable<RawEvent> events) {
        int count = 0;
        for (RawEvent event : events) {
            int updated = jdbcTemplate.update(connection -> {
                PreparedStatement ps = connection.prepareStatement(
                    "INSERT INTO raw_events (id, title, url, published_at, sector, source, payload) " +
                    "VALUES (?, ?, ?, ?, ?, ?, ?) ON CONFLICT (id) DO NOTHING"
                );
                ps.setString(1, event.getId());
                ps.setString(2, event.getTitle());
                ps.setString(3, event.getUrl());
                ps.setObject(4, event.getPublishedAt());
                ps.setString(5, event.getSector());
                ps.setString(6, event.getSource());
                ps.setObject(7, json.toJsonb(event.getPayload()));
                return ps;
            });
            if (updated > 0) {
                count += 1;
            }
        }
        return count;
    }

    public List<RawEvent> fetchUnprocessedRawEvents(int limit) {
        return jdbcTemplate.query(
            "SELECT r.* FROM raw_events r LEFT JOIN normalized_events n ON n.raw_event_id::text = r.id " +
            "WHERE n.raw_event_id IS NULL ORDER BY r.published_at DESC LIMIT ?",
            rawEventMapper(),
            limit
        );
    }

    public RawEvent fetchRawEvent(String rawEventId) {
        List<RawEvent> rows = jdbcTemplate.query(
            "SELECT * FROM raw_events WHERE id = ?",
            rawEventMapper(),
            rawEventId
        );
        return rows.isEmpty() ? null : rows.get(0);
    }

    public void saveNormalized(NormalizedEvent event) {
        jdbcTemplate.update(connection -> {
            PreparedStatement ps = connection.prepareStatement(
                "INSERT INTO normalized_events " +
                "(raw_event_id, event_type, policy_domain, risk_signal, rate_signal, geo_signal, sector_impacts, " +
                "sentiment, rationale, channels, confidence, regime, baseline) " +
                "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?) " +
                "ON CONFLICT (raw_event_id) DO UPDATE SET " +
                "event_type = EXCLUDED.event_type, policy_domain = EXCLUDED.policy_domain, " +
                "risk_signal = EXCLUDED.risk_signal, rate_signal = EXCLUDED.rate_signal, geo_signal = EXCLUDED.geo_signal, " +
                "sector_impacts = EXCLUDED.sector_impacts, sentiment = EXCLUDED.sentiment, rationale = EXCLUDED.rationale, " +
                "channels = EXCLUDED.channels, confidence = EXCLUDED.confidence, regime = EXCLUDED.regime, " +
                "baseline = EXCLUDED.baseline"
            );
            ps.setString(1, event.getRawEventId());
            ps.setString(2, event.getEventType());
            ps.setString(3, event.getPolicyDomain());
            ps.setString(4, event.getRiskSignal());
            ps.setString(5, event.getRateSignal());
            ps.setString(6, event.getGeoSignal());
            ps.setObject(7, json.toJsonb(event.getSectorImpacts()));
            ps.setString(8, event.getSentiment());
            ps.setString(9, event.getRationale());
            ps.setObject(10, json.toJsonb(event.getChannels()));
            ps.setDouble(11, event.getConfidence());
            ps.setObject(12, json.toJsonb(event.getRegime()));
            ps.setObject(13, json.toJsonb(event.getBaseline()));
            return ps;
        });
    }

    public List<NormalizedEvent> fetchUnscoredEvents(int limit) {
        return jdbcTemplate.query(
            "SELECT n.* FROM normalized_events n LEFT JOIN scored_events s ON s.raw_event_id::text = n.raw_event_id::text " +
            "WHERE s.raw_event_id IS NULL ORDER BY n.raw_event_id DESC LIMIT ?",
            normalizedMapper(),
            limit
        );
    }

    public NormalizedEvent fetchNormalizedEvent(String rawEventId) {
        List<NormalizedEvent> rows = jdbcTemplate.query(
            "SELECT * FROM normalized_events WHERE raw_event_id = ?",
            normalizedMapper(),
            rawEventId
        );
        return rows.isEmpty() ? null : rows.get(0);
    }

    public void saveScored(ScoredEvent event) {
        jdbcTemplate.update(connection -> {
            PreparedStatement ps = connection.prepareStatement(
                "INSERT INTO scored_events " +
                "(raw_event_id, event_type, policy_domain, risk_signal, rate_signal, geo_signal, sector_impacts, " +
                "sentiment, rationale, fx_state, sector_scores, total_score, created_at, channels, confidence, regime, baseline) " +
                "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?) " +
                "ON CONFLICT (raw_event_id) DO UPDATE SET " +
                "event_type = EXCLUDED.event_type, policy_domain = EXCLUDED.policy_domain, " +
                "risk_signal = EXCLUDED.risk_signal, rate_signal = EXCLUDED.rate_signal, geo_signal = EXCLUDED.geo_signal, " +
                "sector_impacts = EXCLUDED.sector_impacts, sentiment = EXCLUDED.sentiment, rationale = EXCLUDED.rationale, " +
                "fx_state = EXCLUDED.fx_state, sector_scores = EXCLUDED.sector_scores, total_score = EXCLUDED.total_score, " +
                "created_at = EXCLUDED.created_at, channels = EXCLUDED.channels, confidence = EXCLUDED.confidence, " +
                "regime = EXCLUDED.regime, baseline = EXCLUDED.baseline"
            );
            ps.setString(1, event.getRawEventId());
            ps.setString(2, event.getEventType());
            ps.setString(3, event.getPolicyDomain());
            ps.setString(4, event.getRiskSignal());
            ps.setString(5, event.getRateSignal());
            ps.setString(6, event.getGeoSignal());
            ps.setObject(7, json.toJsonb(event.getSectorImpacts()));
            ps.setString(8, event.getSentiment());
            ps.setString(9, event.getRationale());
            ps.setString(10, event.getFxState());
            ps.setObject(11, json.toJsonb(event.getSectorScores()));
            ps.setDouble(12, event.getTotalScore());
            ps.setObject(13, event.getCreatedAt());
            ps.setObject(14, json.toJsonb(event.getChannels()));
            ps.setDouble(15, event.getConfidence());
            ps.setObject(16, json.toJsonb(event.getRegime()));
            ps.setObject(17, json.toJsonb(event.getBaseline()));
            return ps;
        });
    }

    public ScoredEvent fetchScoredEvent(String rawEventId) {
        List<ScoredEvent> rows = jdbcTemplate.query(
            "SELECT * FROM scored_events WHERE raw_event_id = ?",
            scoredMapper(),
            rawEventId
        );
        return rows.isEmpty() ? null : rows.get(0);
    }

    public List<Map<String, Object>> listTimeline(int limit) {
        return jdbcTemplate.query(
            "SELECT r.title, r.url, r.published_at, r.sector, s.risk_signal, s.rate_signal, s.geo_signal, s.fx_state, " +
            "s.sentiment, s.total_score FROM raw_events r JOIN scored_events s ON s.raw_event_id::text = r.id " +
            "ORDER BY r.published_at DESC LIMIT ?",
            (rs, rowNum) -> {
                Map<String, Object> row = new LinkedHashMap<>();
                row.put("title", rs.getString("title"));
                row.put("url", rs.getString("url"));
                OffsetDateTime published = rs.getObject("published_at", OffsetDateTime.class);
                row.put("published_at", published != null ? published.toString() : null);
                row.put("sector", rs.getString("sector"));
                row.put("risk_signal", rs.getString("risk_signal"));
                row.put("rate_signal", rs.getString("rate_signal"));
                row.put("geo_signal", rs.getString("geo_signal"));
                row.put("fx_state", rs.getString("fx_state"));
                row.put("sentiment", rs.getString("sentiment"));
                row.put("total_score", rs.getDouble("total_score"));
                return row;
            },
            limit
        );
    }

    public Map<String, Double> sectorHeatmap() {
        List<Map<String, Double>> scores = jdbcTemplate.query(
            "SELECT sector_scores FROM scored_events",
            (rs, rowNum) -> json.readDoubleMap(rs.getObject("sector_scores"))
        );

        Map<String, Double> totals = new LinkedHashMap<>();
        for (Map<String, Double> scoreMap : scores) {
            for (Map.Entry<String, Double> entry : scoreMap.entrySet()) {
                totals.put(entry.getKey(), totals.getOrDefault(entry.getKey(), 0.0) + entry.getValue());
            }
        }
        for (String sector : Weights.ALL_SECTORS) {
            totals.putIfAbsent(sector, 0.0);
        }
        Map<String, Double> rounded = new LinkedHashMap<>();
        for (Map.Entry<String, Double> entry : totals.entrySet()) {
            rounded.put(entry.getKey(), Math.round(entry.getValue() * 1000.0) / 1000.0);
        }
        return rounded;
    }

    public List<Map<String, Object>> graphEdges(int limit) {
        List<Map<String, Object>> rows = jdbcTemplate.query(
            "SELECT r.title, s.fx_state, s.risk_signal, s.rate_signal, s.geo_signal, s.sector_scores " +
            "FROM raw_events r JOIN scored_events s ON s.raw_event_id::text = r.id " +
            "ORDER BY r.published_at DESC LIMIT ?",
            (rs, rowNum) -> {
                Map<String, Object> row = new LinkedHashMap<>();
                row.put("title", rs.getString("title"));
                row.put("fx_state", rs.getString("fx_state"));
                row.put("risk_signal", rs.getString("risk_signal"));
                row.put("rate_signal", rs.getString("rate_signal"));
                row.put("geo_signal", rs.getString("geo_signal"));
                row.put("sector_scores", json.readDoubleMap(rs.getObject("sector_scores")));
                return row;
            },
            limit
        );

        List<Map<String, Object>> edges = new ArrayList<>();
        for (Map<String, Object> row : rows) {
            @SuppressWarnings("unchecked")
            Map<String, Double> sectorScores = (Map<String, Double>) row.get("sector_scores");
            for (Map.Entry<String, Double> entry : sectorScores.entrySet()) {
                Map<String, Object> edge = new LinkedHashMap<>();
                edge.put("event", row.get("title"));
                edge.put("fx", row.get("fx_state"));
                edge.put("risk_signal", row.get("risk_signal"));
                edge.put("rate_signal", row.get("rate_signal"));
                edge.put("geo_signal", row.get("geo_signal"));
                edge.put("sector", entry.getKey());
                edge.put("weight", entry.getValue());
                edge.put("fx_theme", row.get("fx_state"));
                edges.add(edge);
            }
        }

        if (edges.isEmpty()) {
            Map<String, Object> fallback = new LinkedHashMap<>();
            fallback.put("event", "Sample event (default)");
            fallback.put("fx", "USD:+0 JPY:+0 EUR:+0 EM:+0");
            fallback.put("risk_signal", "neutral");
            fallback.put("rate_signal", "none");
            fallback.put("geo_signal", "none");
            fallback.put("sector", "Energy");
            fallback.put("weight", 1);
            fallback.put("fx_theme", "neutral");
            edges.add(fallback);
        }

        return edges;
    }

    private RowMapper<RawEvent> rawEventMapper() {
        return (rs, rowNum) -> {
            RawEvent event = new RawEvent();
            event.setId(rs.getString("id"));
            event.setTitle(rs.getString("title"));
            event.setUrl(rs.getString("url"));
            event.setPublishedAt(rs.getObject("published_at", OffsetDateTime.class));
            event.setSector(rs.getString("sector"));
            event.setSource(rs.getString("source"));
            event.setPayload(json.readMap(rs.getObject("payload")));
            return event;
        };
    }

    private RowMapper<NormalizedEvent> normalizedMapper() {
        return (rs, rowNum) -> {
            NormalizedEvent event = new NormalizedEvent();
            event.setRawEventId(rs.getString("raw_event_id"));
            event.setEventType(rs.getString("event_type"));
            event.setPolicyDomain(rs.getString("policy_domain"));
            event.setRiskSignal(rs.getString("risk_signal"));
            event.setRateSignal(rs.getString("rate_signal"));
            event.setGeoSignal(rs.getString("geo_signal"));
            event.setSectorImpacts(json.readIntMap(rs.getObject("sector_impacts")));
            event.setSentiment(rs.getString("sentiment"));
            event.setRationale(rs.getString("rationale"));
            event.setChannels(new ArrayList<>(json.readStringList(rs.getObject("channels"))));
            event.setConfidence(rs.getDouble("confidence"));
            event.setRegime(json.readStringMap(rs.getObject("regime")));
            event.setBaseline(json.readDoubleMap(rs.getObject("baseline")));
            return event;
        };
    }

    private RowMapper<ScoredEvent> scoredMapper() {
        return (rs, rowNum) -> {
            ScoredEvent event = new ScoredEvent();
            event.setRawEventId(rs.getString("raw_event_id"));
            event.setEventType(rs.getString("event_type"));
            event.setPolicyDomain(rs.getString("policy_domain"));
            event.setRiskSignal(rs.getString("risk_signal"));
            event.setRateSignal(rs.getString("rate_signal"));
            event.setGeoSignal(rs.getString("geo_signal"));
            event.setSectorImpacts(json.readIntMap(rs.getObject("sector_impacts")));
            event.setSentiment(rs.getString("sentiment"));
            event.setRationale(rs.getString("rationale"));
            event.setFxState(rs.getString("fx_state"));
            event.setSectorScores(json.readDoubleMap(rs.getObject("sector_scores")));
            event.setTotalScore(rs.getDouble("total_score"));
            event.setCreatedAt(rs.getObject("created_at", OffsetDateTime.class));
            event.setChannels(new ArrayList<>(json.readStringList(rs.getObject("channels"))));
            event.setConfidence(rs.getDouble("confidence"));
            event.setRegime(json.readStringMap(rs.getObject("regime")));
            event.setBaseline(json.readDoubleMap(rs.getObject("baseline")));
            return event;
        };
    }
}
